compiler/IREmitter/IREmitter.cc

```diff
@@ -824,6 +791,81 @@ void IREmitter::run(CompilerState &cs, cfg::CFG &cfg, const ast::MethodDef &md)
         ENFORCE(false);
     }
     cs.runCheapOptimizations(func);
+
+    // If we are ever returning across blocks, we need to wrap the entire execution
+    // of the function in an unwind-protect region that knows about the return.
+    if (!irctx.hasReturnFromBlock) {
+        return;
+    }
+
+    llvm::ValueToValueMapTy VMap;
+    auto *implementationFunction = llvm::CloneFunction(func, VMap);
+
+    // Completely delete the function body.
+    func->dropAllReferences();
+
+    // Turn the original function into a trampoline for this new function.
+    auto *entryBlock = llvm::BasicBlock::Create(cs, "entry", func);
+    builder.SetInsertPoint(entryBlock);
+
+    auto *wrapper = cs.getFunction("sorbet_vm_return_from_block_wrapper");
+    // Adding the function argument at the end means that we don't have to perform
+    // any register shuffling.
```

Comment:
I meant that if you had the four initial arguments passed in registers A, B, C, D, and you, say, passed the function pointer argument first, you'd need to execute, pseudo-code-ly in intel asm syntax:

```
mov E, D
mov D, C
mov C, B
mov B, A,
mov A, function-pointer
```

which is a lot more code than:

```
mov E, function-pointer
```

if you pass the function last.  (Possibly not any slower, since the CPU just does magic register renaming internally, or something?)